### PR TITLE
migrate from Apache HttpClient 4 to Apache HttpClient 5

### PR DESCRIPTION
**Pull request by sullis**
https://hc.apache.org/httpcomponents-client-5.3.x/migration-guide/index.html

https://hc.apache.org/httpcomponents-client-5.3.x/migration-guide/migration-to-classic.html

### CODE REVIEW

File: pinot-common/pom.xml [hunk 1]
```diff
@@ -161,6 +161,10 @@
       <groupId>org.apache.httpcomponents</groupId>
       <artifactId>httpmime</artifactId>
     </dependency>
+    <dependency>
```
Comment (Jackie-Jiang) [inline pinot-common/pom.xml:164]: Thanks for the great effort of this migration. After introducing httpclient5, do we still need to keep the dependency for http4 libraries?

**Comment by sullis** [comment]
ready for review @Jackie-Jiang

**Comment by sullis** [comment]
GitHub Actions:  All checks have passed.

**Comment by sullis** [comment]
Rebased.

**Comment by sullis** [comment]
Rebased.

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MinionClient.java [hunk 2]
```diff
@@ -72,8 +72,8 @@ public Map<String, String> scheduleMinionTasks(@Nullable String taskType, @Nulla
       throws IOException, HttpException {
     HttpPost httpPost = createHttpPostRequest(
         MinionRequestURLBuilder.baseUrl(_controllerUrl).forTaskSchedule(taskType, tableNameWithType));
-    HttpResponse response = HTTP_CLIENT.execute(httpPost);
-    int statusCode = response.getStatusLine().getStatusCode();
+    CloseableHttpResponse response = HTTP_CLIENT.execute(httpPost);
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/minion/MinionClient.java:75]: Do we need to close the response (wrap it in try-with-resource block)? Same for other places

File: pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java [hunk 3]
```diff
@@ -77,1169 +77,1169 @@
  */
 @SuppressWarnings("unused")
 public class FileUploadDownloadClient implements AutoCloseable {
-  private static final Logger LOGGER = LoggerFactory.getLogger(FileUploadDownloadClient.class);
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileUploadDownloadClient.class);
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:80]: The indentation is incorrect, which causes the whole file as changed

File: pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java [hunk 4]
```diff
@@ -76,465 +80,494 @@
...
-    }
-    return requestBuilder.build();
-  }
-
-  @Override
-  public void close()
-      throws IOException {
-    _httpClient.close();
-  }
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpClient.class);
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java:83]: Same here, indentation is changed

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotRunningQueryResource.java [hunk 5]
```diff
@@ -112,50 +116,46 @@ public String cancelQuery(
...
-              .entity(String.format("Query: %s not found on the broker: %s", queryId, brokerId)).build());
-        }
-        throw new WebApplicationException(Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(String
-            .format("Failed to cancel query: %s on the broker: %s with unexpected status=%d and resp='%s'", queryId,
-                brokerId, status, responseContent)).build());
-      } finally {
-        deleteMethod.releaseConnection();
+      Map<String, String> requestHeaders = createRequestHeaders(httpHeaders);
+      requestHeaders.forEach(deleteMethod::setHeader);
+      CloseableHttpResponse response = client.execute(deleteMethod);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotRunningQueryResource.java:140]: We want to close the response

File: pinot-plugins/pinot-environment/pinot-azure/src/main/java/org/apache/pinot/plugin/provider/AzureHttpRequestRetryStrategy.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.plugin.provider;
+
+import org.apache.hc.client5.http.impl.DefaultHttpRequestRetryStrategy;
+import org.apache.hc.core5.util.TimeValue;
+
+
+public class AzureHttpRequestRetryStrategy extends DefaultHttpRequestRetryStrategy {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-environment/pinot-azure/src/main/java/org/apache/pinot/plugin/provider/AzureHttpRequestRetryStrategy.java:25]: Suggest not introducing this class. Directly create a `DefaultHttpRequestRetryStrategy` is more readable

**Review by Jackie-Jiang — COMMENTED** [review]
Thanks for taking time to do this upgrade!
Do we still need to keep httpclient 4 related dependencies?

File: pinot-common/src/main/java/org/apache/pinot/common/minion/MinionClient.java (hunk 2 above)
Comment (sullis) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/MinionClient.java]: Fixed MinionClient.java -- next, I will review other places

Comment (sullis) [reply pinot-common/src/main/java/org/apache/pinot/common/minion/MinionClient.java]: I went through the entire pull request. We are now closing the CloseableHttpResponse instances.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java (hunk 4 above)
Comment (sullis) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java]: Fixed

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotRunningQueryResource.java (hunk 5 above)
Comment (sullis) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotRunningQueryResource.java]: Fixed

File: pinot-plugins/pinot-environment/pinot-azure/src/main/java/org/apache/pinot/plugin/provider/AzureHttpRequestRetryStrategy.java (hunk 6 above)
Comment (sullis) [reply pinot-plugins/pinot-environment/pinot-azure/src/main/java/org/apache/pinot/plugin/provider/AzureHttpRequestRetryStrategy.java]: Fixed.   I removed AzureHttpRequestRetryStrategy.java

File: pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java (hunk 3 above)
Comment (sullis) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java]: Fixed.

**Comment by sullis** [comment]
> Thanks for taking time to do this upgrade! Do we still need to keep httpclient 4 related dependencies?

For now, I think we need to keep httpclient 4 in the project's root pom.xml

I tried removing HttpClient4 and it caused dependency conflicts.  Apache HttpClient 4 is a transitive dependency across multiple libraries:  AWS SDK 2.x Apache Client,  google-api-client, and helix-core

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**